js_modules/dagster-ui/packages/ui-core/src/asset-data/AssetLiveDataProvider.tsx

```diff
@@ -0,0 +1,292 @@
+const BATCHING_INTERVAL = 100;
+
+export const SUBSCRIPTION_IDLE_POLL_RATE = 30 * 1000;
+// const SUBSCRIPTION_MAX_POLL_RATE = 2 * 1000;
+
+type DataForNodeListener = (stringKey: string, data: LiveDataForNode) => void;
+
+const AssetLiveDataContext = React.createContext<{
+  useAssetNodeLiveData: (keys: AssetKeyInput[]) => Record<string, LiveDataForNode | null>;
+}>({
+  useAssetNodeLiveData: () => ({}),
+});
+
+// Map of asset keys to their last fetched time and last requested time
+const lastFetchedOrRequested: Record<
+  string,
+  {fetched: number; requested?: undefined} | {requested: number; fetched?: undefined} | null
+> = {};
+
+export const AssetLiveDataProvider = ({children}: {children: React.ReactNode}) => {
+  const [needsImmediateFetch, setNeedsImmediateFetch] = React.useState<boolean>(false);
+
+  const client = useApolloClient();
+
+  function useAssetNodeLiveData(assetKeys: AssetKeyInput[]) {
+    const [data, setData] = React.useState<Record<string, LiveDataForNode | null>>({});
+
+    React.useEffect(() => {
+      const setDataSingle = (stringKey: string, assetData: LiveDataForNode) => {
+        setData((data) => {
+          return {...data, [stringKey]: assetData};
+        });
+      };
+      assetKeys.forEach((key) => {
+        _subscribeToAssetKey(client, key, setDataSingle, setNeedsImmediateFetch);
+      });
+      return () => {
+        assetKeys.forEach((key) => {
+          _unsubscribeToAssetKey(key, setDataSingle);
+        });
+      };
+    }, [assetKeys]);
+
+    return data;
+  }
+
+  const isDocumentVisible = useDocumentVisibility();
+
+  React.useEffect(() => {
+    if (!isDocumentVisible) {
+      return;
+    }
+    const interval = setInterval(() => {
+      _batchedQueryAssets(_determineAssetsToFetch(), client, (data) => {
+        Object.entries(data).forEach(([key, assetData]) => {
+          const listeners = _assetKeyListeners[key];
+          if (!listeners) {
+            return;
+          }
+          listeners.forEach((listener) => {
+            listener(key, assetData);
+          });
+        });
+      });
+      // Check for assets to fetch every 5 seconds to simplify logic
+      // This means assets will be fetched at most 5 + SUBSCRIPTION_IDLE_POLL_RATE after their first fetch
+      // but then will be fetched every SUBSCRIPTION_IDLE_POLL_RATE after that
+    }, 5000);
+    return () => {
+      clearInterval(interval);
+    };
+  }, [client, isDocumentVisible]);
+
+  React.useEffect(() => {
+    if (!needsImmediateFetch) {
+      return;
+    }
+    const timeout = setTimeout(() => {
+      _batchedQueryAssets(_determineAssetsToFetch(), client, (data) => {
+        Object.entries(data).forEach(([key, assetData]) => {
+          const listeners = _assetKeyListeners[key];
+          if (!listeners) {
+            return;
+          }
+          listeners.forEach((listener) => {
+            listener(key, assetData);
+          });
+        });
+      });
```

Comment:
Would be nice to break this out into a useEffect to make it more clear that it's identical to the code in the hook above

Reply:
Yep will do